Zeroconf review PR

stephanemagnenat opened the pull request:
This goal of this PR is to provide a place for me to comment the newly-created zeroconf feature.

A preliminary question: you wrote feature in a branch that is based on your master branch, I guess it will require some git surgery to merge it into http-refactor, which is based on the aseba-community master branch, right?

davidjsherman committed:
create Aseba::Zeroconf, which announces targets using mDNS-SD

davidjsherman committed:
dummynode target announces itself using Aseba::Zeroconf

davidjsherman committed:
when asking for ephemeral port, don't add deltaPort

stephanemagnenat commented on common/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ set (ASEBACOMMON_SRC
 	utils/BootloaderInterface.cpp
 	msg/msg.cpp
 	msg/NodesManager.cpp
+	zeroconf/zeroconf.cpp
```
There is a missing check for `dns_sd.h` and the related system-dependent library to link.

davidjsherman commented:
I am testing Zeroconf with dummynode only so there should be no conflict with http. The branch is based on aseba-community/master, yes.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+        {
+            DNSServiceErrorType err = DNSServiceProcessResult(z.serviceref); // block until daemon replies
+            if (err != kDNSServiceErr_NoError)
+                throw runtime_error(FormatableString("Zeroconf DNSServiceProcessResult: error %0").arg(err));
```
I would define a subclass of `std::runtime_error` as a public subclass of zeroconf to allow user code to catch this exception specifically.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+	using namespace Dashel;
+
+	//! Register and announce a target described by a name and a port
+	void Zeroconf::registerPort(const string name, const unsigned port, const string txt)
```
The strings should be passed as const references (for example `const string& name`).

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+	}
+
+    //! Register and announce a target described by an existing Dashel stream
+    void Zeroconf::registerStream(const std::string name, const Dashel::Stream* stream, const string txt)
```
The strings should be passed as const reference.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+    //! Register and announce a target described by an existing Dashel stream
+    void Zeroconf::registerStream(const std::string name, const Dashel::Stream* stream, const string txt)
+    {
+        string service_name = name;
```
Why this copy?

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegisterReply: error %0").arg(errorCode));
+        else
+        {
+            ZeroconfService *zref = (ZeroconfService *)context;
```
`Aseba::polymorphic_downcast<T>(base)` (defined in `common/utils/utils.h`) should be used instead of C-style cast.

davidjsherman replied:
Void isn't a class. `DNSServiceRegister` sends an "application context pointer" of our choosing to the callback, it could be NULL, but here I use it to identify the `ZeroconfService` instance that needs to be updated with the name that was actually chosen by the OS—the default behavior of `DNSServiceRegister` is to add a numerical suffix to a name when a conflict occurs e.g. "dummynode-0 (2)".

Alternative 1) Disallow renaming: an attempt to register a Zeroconf name that has already been used will throw an exception. This is probably not a problem, since the proposal in #543 includes the node id in the name, e.g. "Thymio Wireless 5036".

Alternative 2) Maintain a `DNSServiceRef` to `ZeroconfService*` map in the `Zeroconf` instance.

(The updates to `regtype` and `domain` should be removed because these values don't change.)

stephanemagnenat replied:
Oh sorry, I missed that it was a void pointer, my mistake.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    //! Format target info for DNS TXT record
+    string Zeroconf::txtRecord(int protovers, std::string type, std::vector<int> ids, std::vector<int> pids)
```
vectors should be passed as const reference.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+        if (type.size() > 20)
+            type = type.substr(0,20);
+
+        txt.put(9);
```
It would be safer to first put the text to be sent in a string, and then get its length, for instance:

```
string value;
...
value = "txtvers=1";
txt.put(value.length());
txt << value;
```

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+
+        txt.put(4 + 2 * ids.size());
+        txt << "ids=";
+        for (const auto &id : ids) txt.put(id<<8), txt.put(id % 0xff);
```
I would put the body of this loop on an indented new line.

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -0,0 +1,130 @@
+
+        txt.put(5 + 2 * pids.size());
+        txt << "pids=";
+        for (const auto &pid : pids) txt.put(pid<<8), txt.put(pid % 0xff);
```
I would put the body of this loop on an indented new line.

stephanemagnenat commented on common/zeroconf/zeroconf.h:
```diff
@@ -0,0 +1,97 @@
+            ZeroconfService(const std::string name, const std::string domain) : name(name), domain(domain) {}
+            ~ZeroconfService()
+            {   // release the service reference
+                if (serviceref)
```
The DNS Service Discovery C would leave this value uninitialized, thus it could theoretically take any value, so I guess we should initialize it to zero for this test to work.

stephanemagnenat commented on common/zeroconf/zeroconf.h:
```diff
@@ -0,0 +1,97 @@
+            DNSServiceRef serviceref;
+        public:
+            ZeroconfService() {}
+            ZeroconfService(const std::string name, const std::string domain) : name(name), domain(domain) {}
```
This constructor and the previous one are not required in C++11 for the initialize list to work. Yet if we want to initialize `serviceref` to `nullptr` it might be needed.

davidjsherman replied:
OK for removing the constructors and using

```
ZeroconfService z{name,"_aseba._tcp.","local."};
```

for the first three members. Since `serviceref` is a pointer I would expect it to be zero-initialized in the default constructor.

stephanemagnenat replied:
You are right, I was not aware that actually C++ is zero-initializing POD members.

stephanemagnenat commented on common/zeroconf/zeroconf.h:
```diff
@@ -0,0 +1,97 @@
+
+    public:
+        typedef std::map<std::string,std::string> ServiceInfo;
+        std::map<std::string, ServiceInfo> targets;
```
I guess they are only used for target browsing, right? I am not sure whether this should not go to another class or a subclass, as most "light" Aseba programs (dummy, switch) will only advertise services. But it's true that if this structure is empty its memory footprint is low.

stephanemagnenat commented on targets/dummy/dummynode.cpp:
```diff
@@ -82,29 +83,33 @@ class AsebaNode: public Dashel::Hub
 	}
 	
-	void listen(int basePort, int deltaPort)
+    Dashel::Stream* listen(int basePort, int deltaPort)
```
Indentation problem, it is using space rather than tabs corresponding to 4 spaces each.

stephanemagnenat commented on common/zeroconf/zeroconf.h:
```diff
@@ -0,0 +1,97 @@
+        void registerPort(const std::string name, const unsigned port, const std::string txt);
+        void registerStream(const std::string name, const Dashel::Stream* dashel_stream, const std::string txt);
+        void browseForTargets();
+
```
We also need a method to update txt fields of an already-registered server, using `DNSServiceUpdateRecord()` passing `NULL` as the parameter for `RecordRef`.

stephanemagnenat commented:
I've completed my review, I'll be happy to merge this code into http-refactor. This branch is based on your master branch not the community one.

davidjsherman committed:
command-line option --port|-p to request an fixed or an ephemeral port
command-line option --basePort|-b to request an ephemeral port
command-line option --help|-h to print help

davidjsherman committed:
Merge branch 'dummynode-ephemeral' into zeroconf

davidjsherman committed:
use Aseba::Zeroconf

stephanemagnenat commented on targets/dummy/dummynode.cpp:
```diff
@@ -308,20 +312,47 @@ extern "C" void AsebaAssert(AsebaVMState *vm, AsebaAssertReason reason)
+	Dashel::Stream* listen = node.listen(do_delta ? port+deltaNodeId : port, deltaNodeId);
+
+	Aseba::Zeroconf zs;
+	std::string txt_record = Aseba::Zeroconf::txtRecord(ASEBA_PROTOCOL_VERSION, "Unknown", {1+deltaNodeId}, {ASEBA_PID_UNDEFINED});
```
This could be const.

stephanemagnenat replied:
On the one hand, the name could be dummy rather than unknown; on the other hand, its pid is undefined...

davidjsherman committed:
formatting as per Aseba conventions

davidjsherman committed:
Fixes for Zeroconf review PR #32

stephanemagnenat commented on targets/dummy/dummynode.cpp:
```diff
@@ -311,17 +316,33 @@ extern "C" void AsebaAssert(AsebaVMState *vm, AsebaAssertReason reason)
-	}
-	node.listen(basePort, deltaPort);
+
+    int argCounter = 1;
```
wrong indentation

davidjsherman committed:
browsing

TODO: better handling of pending service requests, distinction between service request and record
TODO: allow delegating select loop to Dashel::Hub, QT

davidjsherman committed:
(broken) refactor and delegate event loops

davidjsherman committed:
basic behavior with blocking calls

stephanemagnenat commented on common/zeroconf/zeroconf.cpp:
```diff
@@ -53,15 +53,34 @@ namespace Aseba
 	Zeroconf::Target::Target(const Dashel::Stream* stream)
 	{
 		port = atoi(stream->getTargetParameter("port").c_str());
+		name = "Aseba Local " + stream->getTargetParameter("port");
```
Is there a reason not to use the copy-constructor for port and name, like `port(atoi(...))` and `name("Aseba Local " + ...)`. But it is a detail.

stephanemagnenat commented:
Do you want me to merge it?

davidjsherman commented:
This is the basic version, that blocks on all operations instead of delegating to Dashel::Hub, QT, or std::thread. I haven't yet defined all the virtual methods that will be needed for the derived classes.
There are still two bugs: resolve hangs indefinitely (so doesn't retrieve the port number of discovered nodes), and browsing doesn't reset the tail of the vector that contains discovered (so disconnected nodes aren't forgotten).
There are spurious modifications to dummynode.cpp that shouldn't be in the PR.

stephanemagnenat commented:
This PR is outdated, right?

davidjsherman commented:
Superseded by aseba-community/aseba#554 and aseba-community/aseba#555.
